include/api/wasmedge/wasmedge.hh

```diff
@@ -0,0 +1,707 @@
+#define WASMEDGE_CPP_API_PLUGIN_EXPORT __attribute__((visibility("default")))
+#endif // _WIN32
+
+#include "string"
+#include "vector"
+#include "memory"
+#include "functional"
+
+namespace WasmEdge {
+  enum WASMEDGE_CPP_API_EXPORT ErrCategory {
+    WASM = 0x00,
+    UserLevelError = 0x01
+  };
+
+  enum WASMEDGE_CPP_API_EXPORT HostRegistration {
+    Wasi,
+    WasmEdge_Process,
+    WasiNN,
+    WasiCrypto_Common,
+    WasiCrypto_AsymmetricCommon,
+    WasiCrypto_Kx,
+    WasiCrypto_Signatures,
+    WasiCrypto_Symmetric
+  };
+
+  enum WASMEDGE_CPP_API_EXPORT Proposal {
+    ImportExportMutGlobals = 0,
+    NonTrapFloatToIntConversions,
+    SignExtensionOperators,
+    MultiValue,
+    BulkMemoryOperations,
+    ReferenceTypes,
+    SIMD,
+    TailCall,
+    MultiMemories,
+    Annotations,
+    Memory64,
+    ExceptionHandling,
+    ExtendedConst,
+    Threads,
+    FunctionReferences
+  };
+
+  enum WASMEDGE_CPP_API_EXPORT CompilerOutputFormat {
+    // Native dynamic library format.
+    Native = 0,
+    // WebAssembly with AOT compiled codes in custom sections.
+    Wasm
+  };
+
+  enum WASMEDGE_CPP_API_EXPORT CompilerOptimizationLevel {
+    // Disable as many optimizations as possible.
+    Level_O0 = 0,
+    // Optimize quickly without destroying debuggability.
+    Level_O1,
+    // Optimize for fast execution as much as possible without triggering
+    // significant incremental compile time or code size growth.
+    Level_O2,
+    // Optimize for fast execution as much as possible.
+    Level_O3,
+    // Optimize for small code size as much as possible without triggering
+    // significant incremental compile time or execution time slowdowns.
+    Level_Os,
+    // Optimize for small code size as much as possible.
+    Level_Oz
+  };
+
+  enum WASMEDGE_CPP_API_EXPORT ValType {
+    None = 0x40,
+    I32 = 0x7F,
+    I64 = 0x7E,
+    F32 = 0x7D,
+    F64 = 0x7C,
+    V128 = 0x7B,
+    FuncRef = 0x70,
+    ExternRef = 0x6F
+  };
+
+  enum WASMEDGE_CPP_API_EXPORT ExternalType {
+    Function = 0x00U,
+    Table = 0x01U,
+    Memory = 0x02U,
+    Global = 0x03U
+  };
+
+  enum WASMEDGE_CPP_API_EXPORT Mutability {
+    Const = 0x00,
+    Var = 0x01
+  };
+
+  // TODO: ValType is an extension of RefType
+  enum WASMEDGE_CPP_API_EXPORT RefType {
+    FuncRef = 0x70,
+    ExternRef = 0x6F
+  };
+
+  // TODO: should Version get its own namespace?
+  WASMEDGE_CPP_API_EXPORT std::string VersionGet();
+  WASMEDGE_CPP_API_EXPORT uint32_t VersionGetMajor();
+  WASMEDGE_CPP_API_EXPORT uint32_t VersionGetMinor();
+  WASMEDGE_CPP_API_EXPORT uint32_t VersionGetPatch();
+
+  // TODO: Should Log get its own namespace?
+  WASMEDGE_CPP_API_EXPORT void LogSetErrorLevel();
+  WASMEDGE_CPP_API_EXPORT void LogSetDebugLevel();
+  WASMEDGE_CPP_API_EXPORT void LogOff();
+
+  class WASMEDGE_CPP_API_EXPORT Value {
+  public:
+    // TODO
+  };
+
+  class WASMEDGE_CPP_API_EXPORT String {
+  public:
```

Comment:
I don't think the `String` class is necessary.
In C API, we needs the buffer and length data, so we created the `WasmEdge_String` struct.
But in C++, we can use `std::string` directly.